### ISSUE TITLE
option to install vendored packages into $GOPATH

### ISSUE THREAD

**Issue by cstrahan** [kardianos/govendor#117]
Would it be reasonable to provide a means of installing the vendored packages into `$GOPATH`? It would be nice for other tools that look for everything on `$GOPATH` (like `gocode`), where having `vendor` and `$GOPATH` get out of sync could be confusing. If there was a way to sync in the opposite direction (`vendor --> $GOPATH`), that would be pretty neat.

**Comment by kardianos** [comment]
ala `govendor restore`

Maybe. Would it copy the files into $GOPATH, or revert the repos in $GOPATH to the specified revision if present. Either seems dangerous to me. Generally govendor is designed to never touch anything outside of your project.

**Comment by sean-** [comment]
For the process of clean-room verification and reproducibility, this would be useful.  I think it's fair to never overwrite something, but this is a reasonably common set of operations:

```sh
$ mkdir /tmp/repro
$ cd /tmp/repro
$ export GOPATH=$PWD
$ go get github.com/foo/project
$ cd src/github.com/foo/project
$ cp -prv vendor/g* $GOPATH/src/
```

If any of the external vendored packages exist and are of different versions, the command should abort and the user should manually copy things around.  Even with a `-force-me-harder` flag to a `govendor restore`, I'd hesitate updating any existing packages.

**Comment by kaey** [comment]
Copying vendored packages into GOPATH is a hack, used until all tools properly support vendoring.

**Comment by influx6** [comment]
Although a useful idea but it can a rabbit hole at times, I think unless this is a feature that is widely needed, the users should be the chief decider and initiator of such actions, and I would prefer the user manually copying the needed files over but I also can see the benefit but am just not so convince that benefits makes sense as part of govendor. We could look into a secondary CLi tool that provides these, but the author has the last say. 🙂

**Comment by kardianos** [comment]
I've held off on adding this feature before. I think I would need a concrete proposal that satisfied a reasonably general use case. But given those, I'm open to considering it.

**Comment by devoid** [comment]
One use case would be to install 3rd party go CLI tools in the local environment from a known (vendored) version. You can already add the source to the vendor directory but there's no easy way to `go install` the program, especially if it also uses the vendor path.

**Comment by kardianos** [comment]
@devoid 
Would `govendor install +vendor,program` work for you to install vendor CLI tools?

**Comment by devoid** [comment]
@kardianos When I run `govendor install +vendor,program`, the go toolchain tries to git clone the vendored CLI tool. One of the goals of vendoring these CLI tools is to avoid calling out to the network once you have a working copy of the codebase.

Note that earlier I was having difficulty getting govendor to add all of the required packages for the CLI tool (the tool itself uses govendor but otherwise is a simple go program). Adding the `-tree` option when cloning the repo fixed this issue.

**Comment by kardianos** [comment]
Okay. Right now the govendor {install, test, etc} commands run `govendor sync` before the command. You comment makes me think I should remove that. Normally it would be a noop, but I'm assuming you have vendored deps before I added the checksum.

As for it not vendoring all the deps. Yikes! I just tried it and you're correct, it doesn't do what it needs to do here. I'll add a new test and fix soon.

I think having an implied `govendor sync` will be surprising. I'm going to remove.
I will add an explicit test for main packages and fix the behavior.  Thank you for taking time to report and look into these issue.

**Comment by kardianos** [comment]
@devoid If you run `govendor fetch github.com/my/path/to/cmd/main` It should now correctly and completely vendor the deps. Would you be willing to confirm?

**Comment by devoid** [comment]
@kardianos First sorry for the delayed reply!
Note that all of my steps here are with the v1.0.0 release (congrats!) of govendor and go1.6. Example packages are made up because I don't have permission to release the software.

Just as a recap, I'm trying to vendor a CLI utility I wrote `devoid/mycmd` inside a larger program `devoid/myProgram`. Both source repositories are checked out in my `$GOPATH` and both use the vendor directory and govendor.

In myProgram's directory::
`govendor fetch -v github.com/devoid/mycmd`
Doing this without adding mycmd to govendor results in no output or changes to vendor.json or the vendor files--well it adds a line to vendor.json: `"rootPath": "github.com/devoid/myProgram"` but otherwise no action.

Maybe you wanted me to add first?
`govendor add github.com/devoid/mycmd`, which is checked out in my `$GOPATH`.

No changes to vendor.json or the contents of the directory. What about with `-n`? No output.
`govendor list` shows only things with either `v`, `l` or `lp` in the output. Very strange...
`govendor list | awk '{ print $2 }' | sort | uniq -c` reports no duplicates (saw that issue a few days ago).

Now none of this behavior is what I'd expect. Resetting to a completely clan GOPATH and only cloning my two repositories and govendor v1.0.0 I get the same behavior. It appears I can no longer add packages to my repository.

I'll note that govendor does appear to check that the repositories are in GOPATH as trying to `govendor add github.com/devoid/bogus` returns an error.

**Comment by kardianos** [comment]
@devoid I'm not sure what's happening here. I suspect it is getting filtered out before it can reach either the add or the fetch handler. For sure a bug. I'll see what a testcase for this looks like if I can isolate.

**Comment by devoid** [comment]
An update on this after pulling the [latest govendor commit](https://github.com/kardianos/govendor/commit/8fda9c5c74e854ac24fb33f37565885eb8e0da56).

1. The craziness I was seeing before (with nothing happening and no errors being reported) is now gone. I am guessing this was resolved by the fixes you made to address #130.

2. `govendor fetch github.com/devoid/mycmd` still prompts to download when that code is already checked out in my GOPATH.

3. `govendor add github.com/devoid/mycmd` does not prompt to download. It adds the main.go but not `github.com/devoid/mycmd/pkg/foo`, which is imported in the main package. Running `go list -json github.com/devoid/mycmd` returns DepsErrors and `Incomplete = true` in the [returned structure](https://golang.org/src/cmd/go/list.go#70). This appears to be intentional on the part of the Go authors but I am looking into it.

4. Next running `govendor add +external` pulls in the dependencies for mycmd. At this point I still cannot install mycmd from here since this is not a valid path in the go tools.

Here having something akin to Godep's "restore" function (copy everything from /vendor back into GOPATH) would be useful.

**Comment by rkulla** [comment]
I agree that restoring into $GOPATH would be a hack, so I would like to see the govendor fetch method work better (I have issues with it either not pulling enough down, or putting too much into vendor.json). 

Please also add documentation on this since it's very confusing and requires a lot of trial and error with `add`, `get` and `fetch`. Thanks. I really think govendor is getting close to perfect.

**Comment by kardianos** [comment]
@rkulla 
> Please also add documentation on this since it's very confusing and requires a lot of trial and error with add, get and fetch.

Would you be willing to suggest wording additions or changes that would help you distinguish between those different commands? General or specific suggestions welcome.

**Comment by jbrook** [comment]
If a developer is doing local development on both on a library and a top level project that uses that library at the same time, then vendoring makes the the development workflow a bit awkward. The [official recommendation](https://github.com/golang/go/issues/16562) for Go seems to be to temporarily rename the vendor folder. It would be really handy to run something like 'govendor restore' before doing that so that the GOPATH has the right versions on it. Does that make sense or is there some other alternative workflow that I am missing?
